sklearn/model_selection/_split.py

```diff
@@ -2680,6 +2692,13 @@ def train_test_split(
 
         train, test = next(cv.split(X=arrays[0], y=stratify))
 
+    xp, is_array_api = get_namespace(*arrays)
+    if is_array_api and not _is_numpy_namespace(xp):
+        # Move train and test indexers to the same namespace and device as the
+        # arrays we are indexing. Assumes that all arrays are on the same device
```

Comment:
Alternatively the `cv.split` method could inspect the namespace and device of `X` and use that, no?

Reply:
I agree it's a bit more of an invasive change and what you did can allow us to move forward without having to touch every since CV splitter, including custom ones written by 3rd party libraries or users.

Reply:
I hadn't thought about the option of modifying the splitter to place the indexing arrays on the same device as `X`.

I think maybe we should do both? My thinking is that moving the indexing arrays is probably cheap, so we can do it in `train_test_split` to allow splitters that don't support the Array API to work. We also want to modify the splitters themselves so that not every user of a splitter needs to have this "move the indexing arrays" logic.

WDYT?

Reply:
Something like the following:

```python
def ensure_namespace_device(reference, *arrays, xp=None)
    xp, is_array_api = get_namespace(reference, xp=xp)
    if is_array_api:
        device_ = device(reference)
        return [xp.asarray(a, device=device_) for a in arrays]
    return arrays
```

Reply:
Done in: 2c75f6d.